### PR TITLE
BUG: Prevent VOID_copyswapn ignoring strides

### LINKED ISSUE #8210: Memory corruption when sorting structured array with sub-array
This crashes, either with a segmentation fault or a "corrupted double-linked list" abort:

```
import numpy as np
dt = [('val', 'i4', (1,))]
arr = np.empty(1000, dt)
arr[::-1].sort()
```

I got the idea to reverse the sort order from here: http://stackoverflow.com/a/26984520/4323

Sorting like this always seems to fail if the input array is large enough and it has a structured dtype with at least one element which is an array inside (even in the degenerate case above where the sub-array is of length 1).  It even crashes when trying to sort on a regular scalar field--the mere presence of the array-typed column is enough to make it fail.

I've reproduced this with NumPy 1.11.1 on 64-bit Linux with Python 3.5 and also 2.7.

### PR DESCRIPTION
**Pull request by eric-wieser**
Previously, this would cause a segfault in .sort.

Fixes #8210

### CODE REVIEW

**Comment by charris** [comment]
LGTM once the implicit declaration problem is fixed. I've often thought it would be good if we actually declared all functions at the top of the file before gettin around to the definitions.

File: numpy/core/src/multiarray/arraytypes.c.src [hunk 1]
```diff
@@ -1796,6 +1796,30 @@ BOOL_fromstr(char *str, void *ip, char **endptr,
  */
 
 
+NPY_INLINE void
```
Comment (juliantaylor) [inline numpy/core/src/multiarray/arraytypes.c.src]: needs `static`

**Comment by eric-wieser** [comment]
@juliantaylor: Done

**Comment by juliantaylor** [comment]
can you squash the commits?
I don't like having known broken commits in the history for bisection.

**Comment by eric-wieser** [comment]
@juliantaylor: I'm not sure the first commit is broken? I think the tests just got cancelled

**Comment by eric-wieser** [comment]
It looks like I managed to confuse travis though, as the later tests are definitely running with a different set of source code to the early ones

**Comment by juliantaylor** [comment]
hm looks like it. Its fine then thanks

**Comment by eric-wieser** [comment]
Alarmingly, travis doesn't even have the same sha1 as the commit that tickmark is by

**Pull request merged**